packages/squiggle-lang/src/rescript/Reducer/Reducer_Expression/Reducer_Expression.res

```diff
@@ -38,50 +39,137 @@ let parse_ = (expr: string, parser, converter): result<t, errorValue> =>
+      let rec replaceSymbols = (expression: t, bindings: T.bindings): result<t, errorValue> =>
+        switch expression {
+        | T.EValue(EvSymbol(aSymbol)) =>
+          switch bindings->Belt.Map.String.get(aSymbol) {
+          | Some(boundExpression) => boundExpression->Ok
+          | None => RESymbolNotFound(aSymbol)->Error
+          }
+        | T.EValue(_) => expression->Ok
+        | T.EBindings(_) => expression->Ok
+        | T.EList(list) =>
+          list
+          ->Belt.List.reduceReverse(Ok(list{}), (racc, each: expression) =>
+            racc->Result.flatMap(acc => {
+              each
+              ->replaceSymbols(bindings)
+              ->Result.flatMap(newNode => {
+                acc->Belt.List.add(newNode)->Ok
+              })
+            })
+          )
+          ->Result.map(acc => acc->T.EList)
+        }
+
+      let doBindStatement = (statement: t, bindings: T.bindings) => {
+        switch statement {
+        | T.EList(list{T.EValue(EvCall("$let")), T.EValue(EvSymbol(aSymbol)), expression}) => {
+            let rNewExpression = replaceSymbols(expression, bindings)
+            rNewExpression->Result.map(newExpression =>
+              Belt.Map.String.set(bindings, aSymbol, newExpression)->T.EBindings
+            )
+          }
+        | _ => REAssignmentExpected->Error
+        }
+      }
+
+      let doBindExpression = (expression: t, bindings: T.bindings) => {
+        switch expression {
+        | T.EList(list{T.EValue(EvCall("$let")), ..._}) => REExpressionExpected->Error
+        | _ => replaceSymbols(expression, bindings)
+        }
+      }
+
+      switch list {
+      | list{T.EValue(EvCall("$$bindings"))} => bindings->EBindings->Ok
+
+      | list{T.EValue(EvCall("$$bindStatement")), T.EBindings(bindings), statement} =>
+        doBindStatement(statement, bindings)
+      | list{T.EValue(EvCall("$$bindExpression")), T.EBindings(bindings), expression} =>
+        doBindExpression(expression, bindings)
+      | _ => list->T.EList->Ok
+      }
+    }
+
+    list->dispatchMacroCall(bindings)
   }
 
-/*
-  Recursively evaluate/reduce the code tree
-*/
-let rec reduceExpression = (expression: t, bindings): result<expressionValue, 'e> =>
-  switch expression {
-  | T.EValue(value) => value->Ok
-  | T.EList(list) => {
-      let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
-        racc,
-        each: expression,
-      ) =>
-        racc->Result.flatMap(acc => {
-          each
-          ->reduceExpression(bindings)
-          ->Result.flatMap(newNode => {
-            acc->Belt.List.add(newNode)->Ok
+  let rec seekMacros = (expression: t, bindings: T.bindings): result<t, 'e> =>
+    switch expression {
+    | T.EValue(value) => expression->Ok
+    | T.EList(list) => {
+        let racc: result<list<t>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+          racc,
+          each: expression,
+        ) =>
+          racc->Result.flatMap(acc => {
```

Comment:
what is `Result.flatMap`? shouldn't it be `E.R.bind`?

Reply:
I think Result.flatMap is the same thing as E.R.bind. 

That said, if Result is using ``flatMap`` now, it's probably good for us to change the rest of the codebase to this. (I didn't realize that ``flatMap`` was more standard in Rescript). 

I don't care much for this PR, especially as it's not integrated with E.res yet.

Reply:
I see integration as an issue after clearing major features...

Reply:
ok, resolving then.